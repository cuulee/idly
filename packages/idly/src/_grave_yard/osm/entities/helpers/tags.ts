import { Map } from 'immutable';
export type Tags = Map<string, string>;

export function tagsFactory(obj?: object): Tags {
  return Map(obj);
}

export function osmIsInterestingTag(key) {
  return (
    key !== 'attribution' &&
    key !== 'created_by' &&
    key !== 'source' &&
    key !== 'odbl' &&
    key.indexOf('tiger:') !== 0
  );
}

export const osmOneWayTags = {
  aerialway: {
    chair_lift: true,
    mixed_lift: true,
    't-bar': true,
    'j-bar': true,
    platter: true,
    rope_tow: true,
    magic_carpet: true,
    yes: true
  },
  highway: {
    motorway: true,
    motorway_link: true
  },
  junction: {
    roundabout: true
  },
  man_made: {
    'piste:halfpipe': true
  },
  'piste:type': {
    downhill: true,
    sled: true,
    yes: true
  },
  waterway: {
    canal: true,
    ditch: true,
    drain: true,
    river: true,
    stream: true
  }
};

export const osmPavedTags = {
  surface: {
    paved: true,
    asphalt: true,
    concrete: true
  },
  tracktype: {
    grade1: true
  }
};
