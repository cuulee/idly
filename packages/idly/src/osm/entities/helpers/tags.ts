import { Map } from 'immutable';
export type Tags = Map<string, string>;

export function tagsFactory(obj?: object): Tags {
  return Map(obj);
}
