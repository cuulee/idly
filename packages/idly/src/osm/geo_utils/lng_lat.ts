import { Record } from 'immutable';
export class LngLat extends Record({
  lat: NaN,
  lon: NaN
}) {
  public lat: number;
  public lon: number;
}

export function genLngLat(
  obj: { lon: number; lat: number } | [number, number]
): LngLat {
  if (Array.isArray(obj)) {
    return new LngLat({ lon: obj[0], lat: obj[1] });
  }
  return new LngLat({ ...obj });
}
