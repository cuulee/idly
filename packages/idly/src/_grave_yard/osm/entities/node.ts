import { Record } from 'immutable';

import { Geometry } from 'osm/entities/constants';
import { Properties, propertiesGen } from 'osm/entities/helpers/properties';
import { Tags, tagsFactory } from 'osm/entities/helpers/tags';
import { genLngLat, LngLat } from 'osm/geo_utils/lng_lat';

export type Id = string;
export type Version = number;
export type Visible = boolean;

/**
 * @BUG cant use enum Geometries
 *  Need more investigation, for now
 *  converted it to an object.
 * @REVISIT I think this ^^ bug was due
 *  to cyclic dependencies which I have fixed now.
 *  Enum should in paper work.
 */
console.log('circular dep check', Geometry.POINT);
export class Node extends Record({
  id: 'n-0',
  loc: genLngLat([NaN, NaN]),
  tags: tagsFactory(),
  type: 'node',
  properties: propertiesGen(),
  /**
   * @TOFIX urgent remove this asap
   */
  geometry: Geometry.POINT
}) {
  readonly id: Id;
  readonly tags: Tags;
  readonly type: string;
  readonly loc: LngLat;
  readonly properties: Properties;
  public set(k: string, v: any): Node {
    return super.set(k, v) as Node;
  }
}

export function nodeFactory(
  obj: {
    id: Id;
    tags?: Tags;
    loc?: LngLat;
    properties?: Properties;
  },
  parentWays?
) {
  return new Node(obj);
}
