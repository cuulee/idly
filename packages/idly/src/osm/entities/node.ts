import { List, Record } from 'immutable';

import { genLngLat, LngLat } from 'osm/geo_utils/lng_lat';
import { Properties, propertiesGen } from 'osm/others/properties';
import { ITags, tagsFactory } from 'osm/others/tags';

export type Id = string;
export type Version = number;
export type Visible = boolean;

export class Node extends Record({
  id: 'n-0',
  loc: genLngLat([NaN, NaN]),
  tags: tagsFactory(),
  type: 'node',
  properties: propertiesGen()
}) {
  public id: Id;
  public tags: ITags;
  public type: string;
  public loc: LngLat;
  public properties: Properties;
  public set(k: string, v: any): Node {
    return super.set(k, v) as Node;
  }
}

export function nodeFactory(obj: {
  id: Id;
  tags?: ITags;
  loc?: LngLat;
  properties?: Properties;
}) {
  return new Node(obj);
}
