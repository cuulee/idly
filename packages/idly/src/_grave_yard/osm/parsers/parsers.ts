import { List, Map as ImmutableMap, Set } from 'immutable';

import { Entity } from 'osm/entities/entities';
import { propertiesGen } from 'osm/entities/helpers/properties';
import { tagsFactory } from 'osm/entities/helpers/tags';
import { nodeFactory } from 'osm/entities/node';
import { relationFactory } from 'osm/entities/relation';
import { Way, wayFactory } from 'osm/entities/way';
import { genLngLat } from 'osm/geo_utils/lng_lat';

export type ParentWays = ImmutableMap<string, Set<string>>;

export function calculateParentWays(parentWays: ParentWays, ways: Way[]) {
  console.time('calculateParentWays');
  const x = parentWays.withMutations(p => {
    ways.forEach(w => {
      w.nodes.forEach((n, i) => {
        p.update(n, (s = Set()) => (s = s.add(w.id)));
      });
    });
  });
  console.timeEnd('calculateParentWays');
  return x;
}

export function parseXML(
  xml: XMLDocument,
  parentWays: ParentWays = ImmutableMap()
): {
  entities: Entity[];
  parentWays: ParentWays;
} {
  if (!xml || !xml.childNodes) return;

  const root = xml.childNodes[0];
  const children = root.childNodes;
  const entities: Entity[] = [];
  const nodesXML = [];
  const waysXML = [];
  const relationsXML = [];
  const group = {
    node: [],
    way: [],
    relation: []
  };
  for (let i = 0, l = children.length; i < l; i++) {
    const child = children[i];
    const parser = group[child.nodeName];
    if (parser) group[child.nodeName].push(child);
  }

  group.relation = group.relation.map(parsers.relation);
  group.way = group.way.map(w => parsers.way(w));

  parentWays = calculateParentWays(parentWays, group.way);

  group.node = group.node.map(n => parsers.node(n, parentWays));

  return {
    entities: [...group.node, ...group.way, ...group.relation],
    parentWays
  };
}

function getVisible(attrs) {
  return !attrs.visible || attrs.visible.value !== 'false';
}

function getMembers(obj) {
  const elems = obj.getElementsByTagName('member');
  const members = new Array(elems.length);
  for (let i = 0, l = elems.length; i < l; i++) {
    const attrs = elems[i].attributes;
    members[i] = ImmutableMap({
      id: attrs.type.value[0] + attrs.ref.value,
      type: attrs.type.value,
      role: attrs.role.value
    });
  }
  return List(members);
}

function getNodes(obj) {
  const elems = obj.getElementsByTagName('nd');
  const nodes = new Array(elems.length);
  for (let i = 0, l = elems.length; i < l; i++) {
    nodes[i] = 'n' + elems[i].attributes.ref.value;
  }
  return List(nodes);
}

function getTags(obj) {
  const elems = obj.getElementsByTagName('tag');
  const tags = {};
  for (let i = 0, l = elems.length; i < l; i++) {
    const attrs = elems[i].attributes;
    tags[attrs.k.value] = attrs.v.value;
  }
  return tagsFactory(tags);
}

function getLoc(attrs) {
  const lon = attrs.lon && attrs.lon.value;
  const lat = attrs.lat && attrs.lat.value;
  return genLngLat([parseFloat(lon), parseFloat(lat)]);
}

const parsers = {
  node: function nodeData(obj, parentWays?) {
    const attrs = obj.attributes;
    const id = 'n' + attrs.id.value;
    return nodeFactory({
      id,
      properties: propertiesGen({
        visible: getVisible(attrs),
        version: attrs.version.value,
        timestamp: attrs.timestamp && attrs.timestamp.value,
        changeset: attrs.changeset && attrs.changeset.value,
        uid: attrs.uid && attrs.uid.value,
        user: attrs.user && attrs.user.value
      }),
      loc: getLoc(attrs),
      tags: getTags(obj)
      // geometry: getNodeGeometry(id, parentWays)
    });
  },

  way: function wayData(obj) {
    const attrs = obj.attributes;
    return wayFactory({
      id: 'w' + attrs.id.value,
      properties: propertiesGen({
        visible: getVisible(attrs),
        version: attrs.version.value,
        changeset: attrs.changeset && attrs.changeset.value,
        timestamp: attrs.timestamp && attrs.timestamp.value,
        user: attrs.user && attrs.user.value,
        uid: attrs.uid && attrs.uid.value
      }),
      tags: getTags(obj),
      nodes: getNodes(obj)
    });
    // return way.update('geometry', () => getWayGeometry(way, areaKeys));
  },

  relation: function relationData(obj) {
    const attrs = obj.attributes;
    return relationFactory({
      id: 'r' + attrs.id.value,
      properties: propertiesGen({
        visible: getVisible(attrs),
        version: attrs.version.value,
        changeset: attrs.changeset && attrs.changeset.value,
        timestamp: attrs.timestamp && attrs.timestamp.value,
        user: attrs.user && attrs.user.value,
        uid: attrs.uid && attrs.uid.value
      }),
      tags: getTags(obj),
      members: getMembers(obj)
    });
  }
};
