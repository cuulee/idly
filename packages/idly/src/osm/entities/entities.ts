import { Set } from 'immutable';
import { Node } from 'osm/entities/node';
import { Relation } from 'osm/entities/relation';
import { Way } from 'osm/entities/way';

export type Entities = Set<Node | Way | Relation>;
export type Entity = Node | Way | Relation;
