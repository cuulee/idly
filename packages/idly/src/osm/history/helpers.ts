import { Map } from 'immutable';

import { Graph } from 'osm/history/graph';
import { graphFactory, graphFactoryFromSet } from 'osm/history/graph';

import { Entities, EntitiesId } from 'osm/entities/entities';
import { Node } from 'osm/entities/node';
import { Relation } from 'osm/entities/relation';
import { Way } from 'osm/entities/way';

type EntitiesMap = Map<string, Node | Way | Relation>;
type Entity = Node | Way | Relation;

export function graphSetEntity(graph: Graph, entity: Entity): Graph {
  return graph.setIn([entity.type, entity.id], entity) as Graph;
}

export function graphSetEntitiesSet(graph: Graph, entities: Entities): Graph {
  const patchGraph = graphFactoryFromSet(entities);
  const mergedGraph = graph.mergeWith(
    (oldV, newV) => oldV.merge(newV),
    patchGraph
  ) as Graph;
  return mergedGraph;
}

export function graphSetEntities(graph: Graph, entities: Entity[]): Graph {
  const patchGraph = graphFactory(entities);
  const mergedGraph = graph.mergeWith(
    (oldV, newV) => oldV.merge(newV),
    patchGraph
  ) as Graph;
  return mergedGraph;
}

export function graphRemoveEntities(graph: Graph, entities: Entity[]): Graph {
  return graph.withMutations(g => {
    entities.forEach(e => {
      g.removeIn([e.type, e.id]);
    });
  }) as Graph;
}

export function graphRemoveEntitiesWithId(
  graph: Graph,
  entities: EntitiesId
): Graph {
  return graph.withMutations(g => {
    entities.forEach(e => {
      if (e[0] === 'n') g.removeIn(['node', e]);
      else if (e[0] === 'w') g.removeIn(['way', e]);
      else g.removeIn(['relation', e]);
    });
  }) as Graph;
}

export function graphRemoveEntity(graph: Graph, entity: Entity): Graph {
  return graph.removeIn([entity.type, entity.id]) as Graph;
}

export function calculateParentWays(graph: Graph): any {
  let parentWays = Map();
  parentWays = parentWays.withMutations(map => {
    graph.way.forEach(e => {
      e.nodes.forEach(n => {
        map.set(n, e.id);
      });
    });
  });
  return parentWays;
}
