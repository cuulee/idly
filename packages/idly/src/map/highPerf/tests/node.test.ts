import { fromJS, Map, Set } from 'immutable';

import { getNodeGeometry, nodeToPoint } from 'map/highPerf/node';
import { tagsFactory } from 'osm/entities/helpers/tags';
import { nodeFactory } from 'osm/entities/node';
import { genLngLat } from 'osm/geo_utils/lng_lat';
import { ParentWays } from 'osm/parsers/parsers';

const n1 = nodeFactory({ id: 'n-1' });
const n11 = nodeFactory({
  id: 'n-1',
  loc: genLngLat({ lon: 15, lat: 10 }),
  tags: tagsFactory({ k: 'k' })
});

describe('converts node to feat', () => {
  describe('nodeToPoint', () => {
    it('should behave...', () => {
      expect(nodeToPoint(n1)).toMatchSnapshot();
      expect(nodeToPoint(n11)).toMatchSnapshot();
    });
  });

  describe('getNodeGeometry', () => {
    it('should work for empty parentWays', () => {
      expect(getNodeGeometry('n1', Map()));
    });
  });
});
