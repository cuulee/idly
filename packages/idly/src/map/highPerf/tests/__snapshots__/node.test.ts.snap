// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`converts node to feat nodeToPoint should behave... 1`] = `
Object {
  "geometry": Object {
    "coordinates": Array [
      NaN,
      NaN,
    ],
    "type": "Point",
  },
  "id": "n-1",
  "properties": Object {},
  "type": "Feature",
}
`;

exports[`converts node to feat nodeToPoint should behave... 2`] = `
Object {
  "geometry": Object {
    "coordinates": Array [
      15,
      10,
    ],
    "type": "Point",
  },
  "id": "n-1",
  "properties": Object {},
  "type": "Feature",
}
`;
